spark-3.1-spanner-lib/src/test/java/com/google/cloud/spark/spanner/ReadIntegrationTestBase.java

```diff
@@ -58,39 +60,191 @@ public void testReadFromCompositeTable() {
     List<String> expectIds = Arrays.asList("id1", "id2");
     assertThat(gotIds).containsExactlyElementsIn(expectIds);
 
-    // 2. Validate C field string values.
+    // 2. Validate A field ARRAY<INT64> values.
+    List<Row> allAs = df.select("A").collectAsList();
+    assertEquals(allAs.size(), 2);
+    Row rowA1 = allAs.get(0);
+    assertEquals(rowA1.size(), 1);
+    Row rowA2 = allAs.get(1);
+    assertEquals(rowA2.size(), 1);
+    List<Long> gotAs = new ArrayList();
+    gotAs.addAll(rowA1.getList(0));
+    gotAs.addAll(rowA2.getList(0));
+    Collections.sort(gotAs);
+
+    long[] lA = {10, 100, 991, 567282, 20, 200, 2991, 888885};
+    List<Long> wantAFlattened = Arrays.stream(lA).boxed().collect(Collectors.toList());
+    Collections.sort(wantAFlattened);
+    assertThat(gotAs).containsExactlyElementsIn(wantAFlattened);
+
+    // 3. Validate B field ARRAY<STRING> values.
+    List<Row> allBs = df.select("B").collectAsList();
+    assertEquals(allBs.size(), 2);
+    Row rowB1 = allBs.get(0);
+    assertEquals(rowB1.size(), 1);
+    Row rowB2 = allBs.get(1);
+    assertEquals(rowB2.size(), 1);
+    List<String> gotBs = new ArrayList();
+    gotBs.addAll(rowB1.getList(0));
+    gotBs.addAll(rowB2.getList(0));
+    Collections.sort(gotBs);
+
+    List<String> wantBFlattened = Arrays.asList("a", "b", "c", "A", "B", "C");
+    Collections.sort(wantBFlattened);
+    assertThat(gotBs).containsExactlyElementsIn(wantBFlattened);
+
+    // 4. Validate C field string values.
     List<String> gotCs = df.select("C").as(Encoders.STRING()).collectAsList();
     List<String> expectCs = Arrays.asList("foobar", "this one");
     assertThat(gotCs).containsExactlyElementsIn(expectCs);
 
-    // 3. Validate G field boolean values.
-    List<Boolean> gotGs = df.select("G").as(Encoders.BOOLEAN()).collectAsList();
-    List<Boolean> expectGs = Arrays.asList(true, false);
-    assertThat(gotGs).containsExactlyElementsIn(expectGs);
+    // 5. Validate D field numeric values.
+    List<BigDecimal> gotDs = df.select("D").as(Encoders.DECIMAL()).collectAsList();
+    List<BigDecimal> expectDs =
+        Arrays.asList(asSparkBigDecimal("2934"), asSparkBigDecimal("93411"));
+    Collections.sort(gotDs);
+    Collections.sort(expectDs);
+    assertThat(gotDs).containsExactlyElementsIn(expectDs);
 
-    // 4. Validate E field date values.
+    // 6. Validate E field date values.
     List<Date> gotEs = df.select("E").as(Encoders.DATE()).collectAsList();
     List<Date> expectEs = Arrays.asList(Date.valueOf("2022-12-31"), Date.valueOf("2023-09-22"));
     assertThat(gotEs).containsExactlyElementsIn(expectEs);
 
-    // 5. Validate F field timestamp values.
+    // 7. Validate F field timestamp values.
     List<Timestamp> gotFs = df.select("F").as(Encoders.TIMESTAMP()).collectAsList();
     List<Timestamp> expectFs =
         Arrays.asList(
             new Timestamp(ZonedDateTime.parse("2023-08-26T12:22:05Z").toInstant().toEpochMilli()),
```

Comment:
Can we check the limit of timestamp. I assume we only convert the microseconds from Spanner to spark. Are we going to add such test in the testNumericLimits or testReadFromCompositeTable ?

Reply:
We do so in testValueLimits